PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+			const auto search_result = search_tree(this->root.as_node_ref(), key);
+			switch (search_result.result) {
+				case SearchResult::FOUND:  return Entry::Occupied(search_result.handle, this->len);
+				case SearchResult::FAILED: return Entry::Vacant(  search_result.handle, this->len, key);
+				default: UNREACHABLE();
+			}
+		}
+		Range<K, V, Bfactor> MaxRange() { return {BTree::first_leaf_edge(this->root.as_node_ref()), BTree::last_leaf_edge(this->root.as_node_ref())}; }
+		Range<K, V, Bfactor> GetRange(const K& lower, const K& upper) { ASSERTE(lower < upper, ""); return {search_tree(this->root.as_node_ref(), lower).handle, search_tree(this->root.as_node_ref(), upper).handle}; }
+		Iter begin() { return {BTree::first_leaf_edge(this->root.as_node_ref()), this->GetSize(), this->GetSize()}; }
+		Iter end()   { return {BTree:: last_leaf_edge(this->root.as_node_ref()), 0,               this->GetSize()}; }
+		class Keys   Keys()   { return {*this}; }
+		class Values Values() { return {*this}; }
+
+	public:
+		void swap(BTreeMap& other) { std::swap(this->root, other.root); std::swap(this->len, other.len); }
+
+	private:
+		template<typename Key, typename Val>
+		Optional<V> insert(Key&& key, Val&& value) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (entry.IsVacant()) {
+				entry.VacantInsert(std::forward<Val>(value));
+				return {};
+			} else {
+				return {entry.OccupiedInsert(std::forward<Val>(value))};
+			}
+		}
+
+		template<typename Key>
+		Optional<V> remove(Key&& key) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (!entry.IsVacant()) {
+				return {entry.Remove()};
+			} else {
+				return {};
+			}
+		}
+
+		template<typename Key>
+		Optional<V&> get(Key&& key) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (!entry.IsVacant()) {
+				return {entry.OccupiedGet()};
+			} else {
+				return {};
+			}
+		}
+
+	private:
+		class Entry	{
+		public:
+			static Entry Vacant  (KVERef kv_ref, size_t& len, const K&  key) { return Entry(kv_ref, len, key); }
+			static Entry Vacant  (KVERef kv_ref, size_t& len,       K&& key) { return Entry(kv_ref, len, std::move(key)); }
+			static Entry Occupied(KVERef kv_ref, size_t& len)                { return Entry(kv_ref, len); }
+
+		public:
+			bool IsVacant() { return static_cast<bool>(key); }
+
+			V& OrInsert(const V&  value) { return this->or_insert(value); }
+			V& OrInsert(      V&& value) { return this->or_insert(std::move(value)); }
+
+			V& VacantInsert(const V&  value) { return this->vacant_insert(value); }
+			V& VacantInsert(      V&& value) { return this->vacant_insert(std::move(value)); }
+
+			V OccupiedInsert(const V&  value) { return this->occupied_insert(value); }
+			V OccupiedInsert(      V&& value) { return this->occupied_insert(std::move(value)); }
+
+			V& OccupiedGet() { ASSERTE(!this->IsVacant(), "Cannot read value from a vacant map entry!"); return this->kv_ref.node_ref.node->values[this->kv_ref.idx]; }
```

Comment:
I agree the API could be nicer if we had:
- sum types *with* pattern matching
- affine type system

The first point would address the possibility of calling wrong method on the wrong variant (e.g. `OccupiedGet()` on `VacantEntry`).
The second would take care of the invalidation problem (as entries would get consumed on insert/remove).
With that being said, it's still not _that_ unsafe. The incorrect variant issue is fully defined and safe (just as `Remove` of `Remove`/`TryRemove` combo is); the assertion will fail.
```
collection.TryRemove(value); //always ok
collection.Remove(value); //may be a programmer error
```
```
map.Entry(of_key).OccupiedGet(); //essentially the same
```
(I now realized that I had forgotten to make entries implicitly convertible to boolean values, which would enable the following pattern ⟨as there are only 2 possible variants⟩)
```
if (auto entry = map.Entry(key)) {
	//do stuff with occupied entry
} else {
	//insert
}
```
And then in C++17 one can do the more descriptive
```
if (auto entry = map.Entry(key); !entry.IsVacant())
```
There is nothing however that can be done about invalidation as evidenced by even the standard library iterators (we need proper lifetime enforcement for that)
```
auto entry = map.Entry(key);
entry.Remove();
entry.Remove(); //uh oh
```
```
auto iter = map.find(foo);
map.insert(bar);
*iter; //oh boy
```